DEV: Governance document (draft)

njsmith opened the pull request:
This probably shouldn't be merged yet, but since we seem to be starting to converge and there has been some feedback on specific text, let's get it up here where we can see it.

This is intentionally structured as multiple commits -- the first commit is a close translation of the version that was originally posted to the list, and then I'm making small commits on top of this, because it seems useful to record this history in a way that will let it be reviewed later. (And also, uh, now.)

The main differences from the version originally posted are:
- Converted from markdown to rst
- Moved the lists of names out into a separate "people" page (similar to how https://github.com/jupyter/governance is structured now -- see https://github.com/jupyter/governance/blob/master/people.md)
- Removed the link to the github contributors page (as per discussion at https://mail.scipy.org/pipermail/numpy-discussion/2015-September/073712.html)

Also, I stuck it into the dev docs -- if anyone has a suggestion for a better place let me know.

njsmith committed:
DEV: Draft governance document + list of people and positions

This is definitely *not* the final version -- it's the version
originally posted to the mailing list, reformatted as ReST. I'll make
further changes on top of this as further commits, in order to preserve
the historical record.

njsmith committed:
DEV: Governance: delete redundant bits

The NumFOCUS subcommittee and Institutional Partners are now listed in a
separate file, so delete them from here.

We will eventually want to clean up the list of initial members of the
steering council, but the details are still under discussion on the
mailing list, so I left that section alone for now.

njsmith committed:
DEV: Governance: remove link to inaccurate github contributors page

See Travis's email here:
  https://mail.scipy.org/pipermail/numpy-discussion/2015-September/073712.html

njsmith committed:
DEV: Governance: update link to Jupyter/IPython governance doc

njsmith committed:
DEV: Governance: fix pandoc weirdness

I used pandoc to convert Markdown to ReST, and for some reason it marked
all the links with two underscores instead of the more conventional
single underscore. I'm not sure why.

njsmith commented:
Also, as a suggestion: let's try to keep discussion of any semantic changes to the mailing list for now, so as not to confuse things.

njsmith commented on doc/source/dev/governance/people.rst:
```diff
@@ -0,0 +1,49 @@
+Emeritus members
+----------------
+
+* Travis Oliphant - Project Founder / Emeritus Leader (served: 2001(??)-2012)
```
@teoliphant: Can you check the years here?

njsmith committed:
DEV: minor textual cleanups to governance document

njsmith committed:
DEV: Fix dates on Travis's time as leader

The 2001 date I had originally was based on the beginning of NumPy's git
history -- but further research suggests that this actually runs back
into Numeric times, before the NumPy project was founded.

New 2005 date is based on Fernando's recollection of the fateful meeting
that led to NumPy's creation, and is consistent with Wikipedia's claim
that 2006 was the first release.

If this is wrong please let me know :-)

njsmith committed:
DEV: remove the interstitial notes on the seed steering council

Now that this seems to be settled, we don't need this cluttering up the
main text. (For the record: the rule listed here is what we actually
used.)

njsmith committed:
DEV: remove draft markers from governance document

In preparation for merging.

charris commented on doc/source/dev/governance/governance.rst:
```diff
@@ -0,0 +1,376 @@
+Project Contributor who has produced contributions that are substantial
+in quality and quantity, and sustained over at least one year. Potential
+Council Members are nominated by existing Council members and voted upon
+by the existing Council after asking if the potential Member is
```
Hmm, the order could be reversed here. It might make sense for the vote to precede the invitation so as to avoid potential  embarassment. The vote could be made part of the nomination process.

njsmith replied:
I suspect the reason this got past the IPython folks was that in practice no-one is formally nominated unless there's already consensus that they'd be a good choice, so it wouldn't be an issue in practice. But, updated the text to clarify that these are two conditions that have to be met, rather than that one specifically has to follow the other.

charris commented on doc/source/dev/governance/governance.rst:
```diff
@@ -0,0 +1,376 @@
+by the existing Council after asking if the potential Member is
+interested and willing to serve in that capacity. The Council will be
+initially formed from the set of existing Core Developers who, as of
+late 2015, have been significantly active over the last year.
```
Significantly is too vague, I think. Original attempt was commit bit, second was committed a merge (?), which both had the advantage of being public and concrete.

charris replied:
We have to be able to say who will be on the council, and for subjective qualities we would need to set up a nominating committee, and recurse forever.

charris replied:
OK, I see that names have been named below, which is somewhat different. Should at least be referenced here.

njsmith replied:
Added links to the governance document ("the current steering council membership is here", "the current numfocus subcommittee is here", "the current institutional partners are listed here").

I don't think we need to maintain a permanent record here of all the details of how we bootstrapped the initial council, because it isn't relevant going forward, and if anyone is ever curious about the details they can always be found in the mailing list archives or in git history: https://github.com/njsmith/numpy/commit/29b09f86b7956e822751f7da54a998395a2baf76

charris commented on doc/source/dev/governance/governance.rst:
```diff
@@ -0,0 +1,376 @@
+Private communications of the Council
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Unless specifically required, all Council discussions and activities
```
?. 'Unless specifically required' doesn't fit here. Maybe 'Unless privacy is required'

njsmith replied:
Made it "to the maximum extent possible", which I think works? Notice that the next sentence gives more details on how and when private conversation happens.

charris commented on doc/source/dev/governance/governance.rst:
```diff
@@ -0,0 +1,376 @@
+approve the Partnership.
+
+If an existing Institutional Partner no longer has a contributing
+employee, they will be given a 1 year grace period for remaining
```
This is unclear. I think what we are aiming for here is that employees come an go and as long as the gap between employees does not exceed a year, it is OK. One way to say that is that is that the Institutional Partnership will lapse after a period of one year with no active employee.

njsmith replied:
Yes, that's my interpretation of what we want too.

Tried to rewrite to be more clear -- see what you think.

charris commented:
Looks almost ready ;) Thank Nathaniel.

njsmith committed:
DEV: for new council members, clarify order of voting and talking to them

Original text seemed to imply that we would always ask the potential new
Council Member whether they were interested *before* we actually decided
whether to offer them a spot, which could create a sticky situation if
someone ever got voted down.

Rephrase to make clear that things do not have to occur in this order.

njsmith committed:
DEV: governance: Include links to current membership of each group

njsmith committed:
DEV: governance: wording tweak

njsmith committed:
DEV: attempt to clarify rules for ending an Institutional Partnership

njsmith commented:
Okay, I think I addressed everything!

charris commented:
Nice rewording! Maybe one more. I suggest adding Allan Haldane to the initial Steering Council. He has already merged a PR, has a PR in the works, and is reviewing some of the other PRs.

njsmith commented:
@charris: I've also been impressed by @ahaldane's contributions, but I'm somewhat uncomfortable about tweaking stuff like this at the last minute like this with just the two of us looking, and without having a clear picture of his contributions off the top of my head (has he been around a year already?). Would it be all right to put this in now as it is, and then consider that as the new council's first act?

seberg commented:
Just to note, @rgommers will be back in a couple of days. I do not know whether he looked it closely before he left, but if he did not, maybe we can wait for him.

charris commented:
@njsmith I'd just do it, it's as much diplomacy as anything else. @seberg Waiting for Ralf to take a look sounds reasonable, athough it may not be the first thing he wants to do when getting back ;) But a few days here or there aren't going to matter.

njsmith commented:
> @njsmith I'd just do it, it's as much diplomacy as anything else.

At this point I'm not willing to "just do it" on anything but wording tweaks without broader discussion. I'm sure we can trust @ahaldane to understand that this is simply because the goal of this process is to make these decisions more predictable and explicable to observers, and is not at all a judgement on their contributions :-).

Re: Ralf: Oh, good point, I hadn't realized that we were already that close to Ralf getting back. I'm pretty sure he left just before I finally finished and posted the draft (and ditto for the roadmap stuff), so I would assume he hasn't seen it.

My guess from talking to him before about governance issues is that his reaction will mostly be "oh thank goodness it's finally done". So I'd be comfortable with merging now, and if it turns out that he does have any concerns then it'll be easy enough to fix those even after it's merged. (The criterion for updates to the governance document is just consensus, same as any other change.) But if you all prefer to wait and get his sign-off explicitly then I'm fine with that too.

charris commented:
Well, let's just get this in. However, there should probably be a note somewhere that states when it takes effect. Perhaps on merge?

njsmith commented:
How about we merge it and then one of us sends an email to the list
announcing that as of <whenever the email is sent> it is now in effect?

On Mon, Oct 5, 2015 at 3:23 PM, Charles Harris <notifications@github.com>
wrote:

> Well, let's just get this in. However, there should probably be a note
> somewhere that states when it takes effect. Perhaps on merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6352#issuecomment-145686598>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

ahaldane commented:
FWIW, I am still learning about the ideas for the future of numpy, so I probably wouldn't want to contribute much to that kind of discussion right now. I am sure if it is ever important it can be changed later.

Of course if I ended up on the council I wouldn't object. As Francois Arago once said when advised to turn down nomination to the National Academy, "I don't intend to make myself guilty of inconveniencing the Academy."

I suggest merging this without me on it, though I thank @charris for the consideration!

charris commented:
OK, merged. Thanks Nathaniel. I'll send a note to the list.

njsmith commented:
Woot! Thanks Chuck!

On Mon, Oct 5, 2015 at 4:18 PM, Charles Harris <notifications@github.com>
wrote:

> OK, merged. Thanks Nathaniel. I'll send a note to the list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6352#issuecomment-145696122>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

charris commented:
We now need to set up some infrastructure. The two items that occur to me off the top are

* Private mailing list for the Steering Committee.
* Public site where the current membership of the Steering Commitee is listed, and probably a link to the  Governance Document as well.

njsmith commented:
The mailing list is a bit tricky b/c no-one seems to know who is actually running ou rmailing list infrastructure irght now :-) But see https://github.com/numpy/numpy/issues/6325#issuecomment-145483313

As soon as the docs get rebuilt, we will have a public web site with the steering committee -- the .rst files that were just merged are built into the public docs. I'm not sure if we have an auto-updated version of the docs?

rgommers commented:
Hi all, I'm back amongst the living. This looks excellent, thanks for putting in so much effort! 

Regarding mailing lists:
- the numpy list is administered by Enthought.
- I don't see a reason for why the new steering committee list needs to be administered the same way as the numpy list, especially given the trouble we've had with the latter.
- I'd go for a closed Google Groups list for the steering committee list, with all committee members having admin access to it.
